### DISCUSSION TITLE
Adding the option change Mesh Viewer

### DISCUSSION THREAD

**Discussion by classicrocker883** [mriscoc/Ender3V2S1#411] [Ideas]
### Adding Mesh Viewer - change how the mesh is viewed
enables both options on how the mesh is viewed
with `USE_UBL_VIEWER` enabled, this gives a check box to alternate type of view for mesh

</BR>

</BR>

adding Menu Item
FILE-->   dwin.cpp
   under Draw AdvancedSettings Menu
   

      ```MENU_ITEM(ICON_MeshViewer, MSG_MESH_VIEW, onDrawSubMenu, DWIN_MeshViewer);  
      #if ENABLED(USE_UBL_VIEWER)
        EDIT_ITEM_F(ICON_PrintSize, "Change Mesh Viewer", onDrawChkbMenu, SetViewMesh, &BedLevelTools.view_mesh); 
      #endif```
</BR>

</BR>
      
      
//the function      
FILE-->   meshviewer.cpp

```
void MeshViewerClass::Draw(bool withsave /*= false*/) {
  Title.ShowCaption(GET_TEXT_F(MSG_MESH_VIEWER));
  #if ENABLED(USE_UBL_VIEWER)
  if(BedLevelTools.view_mesh) {
    DWINUI::ClearMainArea();
    BedLevelTools.viewer_print_value = true;
    BedLevelTools.Draw_Bed_Mesh(-1, 1, 8, 10 + TITLE_HEIGHT);}
    else
      DrawMesh(bedlevel.z_values, GRID_MAX_POINTS_X, GRID_MAX_POINTS_Y);
  #else
    DrawMesh(bedlevel.z_values, GRID_MAX_POINTS_X, GRID_MAX_POINTS_Y);
  #endif
  if (withsave) {
    DWINUI::Draw_Button(BTN_Save, 26, 305);
    DWINUI::Draw_Button(BTN_Continue, 146, 305);
    Draw_Select_Highlight(HMI_flag.select_flag, 305);
  }
  else
    DWINUI::Draw_Button(BTN_Continue, 86, 305);

  #if ENABLED(USE_UBL_VIEWER)
    if(BedLevelTools.view_mesh) {
      BedLevelTools.Set_Mesh_Viewer_Status();}
    else {
    char str_1[6], str_2[6] = "";
    ui.status_printf(0, F("Mesh minZ: %s, maxZ: %s"),
      dtostrf(min, 1, 2, str_1),
      dtostrf(max, 1, 2, str_2)
    );}
  #else
    char str_1[6], str_2[6] = "";
    ui.status_printf(0, F("Mesh minZ: %s, maxZ: %s"),
      dtostrf(min, 1, 2, str_1),
      dtostrf(max, 1, 2, str_2)
    );
  #endif
```

 
</BR>

</BR>
  
  
  
  
FILE--> settings.cpp

```
  #if ENABLED(USE_UBL_VIEWER)
    bool view_mesh;
  #endif
```
//under MarlinSettings::save()

```
  #if ENABLED(USE_UBL_VIEWER)
    EEPROM_WRITE(BedLevelTools.view_mesh);
  #endif
```    
//under MarlinSettings:_load()

```
  #if ENABLED(USE_UBL_VIEWER)
    _FIELD_TEST(view_mesh);
    EEPROM_READ(BedLevelTools.view_mesh);
  #endif      
```
//under MarlinSettings:reset() 
```
  #if ENABLED(USE_UBL_VIEWER)
    BedLevelTools.view_mesh = ENABLED(USE_UBL_VIEWER); //added mesh viewer option
  #endif 
```
</BR>

</BR>


FILE--> bedlevel_tools.cpp
 
//under
```BedLevelToolsClass BedLevelTools
 
  #if ENABLED(USE_UBL_VIEWER)
    static bool viewer_asymmetric_range;
    static bool viewer_print_value;
    static bool view_mesh;
  #endif
```
  
</BR>

</BR> 
  
FILE--> bedlevel_tools.h

```
  class BedLevelToolsClass {
public:
  #if ENABLED(USE_UBL_VIEWER)
    static bool viewer_asymmetric_range;
    static bool viewer_print_value;
    static bool view_mesh;
  #endif
```
  

![20220831_011809](https://user-images.githubusercontent.com/18502096/187598211-033c08d0-2201-4b33-b4b8-4a7faa4a49f4.jpg)
![20220831_011731](https://user-images.githubusercontent.com/18502096/187598302-6bffd730-4bc6-40c8-96e7-666dc63b6219.jpg)

**Comment by mriscoc** [comment]
Be aware that this duplicates the code of the viewer, that is why I leave it only as a user's selectable at compile time.
Currently the RCT6 devices are close to the limit of the program's memory.

**Comment by classicrocker883** [comment]
i see, maybe that can be tweaked so it limit the memory

**Comment by classicrocker883** [comment]
using the RET6...

Without `USE_UBL_VIEWER` enabled
RAM:   [==        ]  23.3% (used 15300 bytes from 65536 bytes)
Flash: [====      ]  41.7% (used 218444 bytes from 524288 bytes)

With `USE_UBL_VIEWER` enabled
RAM:   [==        ]  23.4% (used 15304 bytes from 65536 bytes)
Flash: [====      ]  41.9% (used 219876 bytes from 524288 bytes)

**Comment by mriscoc** [comment]
Yes, RET6 doesn't have a problem because it has more program memory than RCT6, but many new printers are now with RCT6 and you could soft-brick those printers.
